rmf_gazebo_plugins/src/TeleportIngestor.cpp

```diff
@@ -0,0 +1,343 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <vector>
+#include <algorithm>
+#include <unordered_map>
+#include <unordered_set>
+
+#include <gazebo/common/Plugin.hh>
+#include <gazebo/physics/Model.hh>
+#include <gazebo/physics/World.hh>
+#include <gazebo/physics/physics.hh>
+#include <gazebo_ros/node.hpp>
+
+#include <ignition/math/Vector3.hh>
+
+#include "utils.hpp"
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <rmf_fleet_msgs/msg/fleet_state.hpp>
+#include <rmf_dispenser_msgs/msg/dispenser_state.hpp>
+#include <rmf_dispenser_msgs/msg/dispenser_result.hpp>
+#include <rmf_dispenser_msgs/msg/dispenser_request.hpp>
+
+namespace rmf_gazebo_plugins {
+
+class TeleportIngestorPlugin : public gazebo::ModelPlugin
+{
+
+public:
+
+  using FleetState = rmf_fleet_msgs::msg::FleetState;
+  using DispenserState = rmf_dispenser_msgs::msg::DispenserState;
+  using DispenserRequest = rmf_dispenser_msgs::msg::DispenserRequest;
+  using DispenserResult = rmf_dispenser_msgs::msg::DispenserResult;
+  using Pose3d = ignition::math::Pose3d;
+
+private:
+
+  std::string _guid;
+  double _last_pub_time = 0.0;
+  bool _load_complete = false;
+
+  gazebo::event::ConnectionPtr _update_connection;
+  gazebo::physics::ModelPtr _model;
+  gazebo::physics::ModelPtr _ingested_model;
+  gazebo::physics::WorldPtr _world;
+
+  gazebo_ros::Node::SharedPtr _node;
+  rclcpp::Subscription<FleetState>::SharedPtr _fleet_state_sub;
+  rclcpp::Publisher<DispenserState>::SharedPtr _state_pub;
+  rclcpp::Subscription<DispenserRequest>::SharedPtr _request_sub;
+  rclcpp::Publisher<DispenserResult>::SharedPtr _result_pub;
+
+  std::unordered_map<std::string, ignition::math::Pose3d> 
+      _non_static_models_init_poses;
+
+  std::unordered_map<std::string, FleetState::UniquePtr> _fleet_states;
+
+  std::unordered_map<std::string, bool> _past_request_guids;
+
+  DispenserState _current_state;
+
+  rclcpp::Time simulation_now() const
+  {
+    const double t = _model->GetWorld()->SimTime().Double();
+    const int32_t t_sec = static_cast<int32_t>(t);
+    const uint32_t t_nsec =
+        static_cast<uint32_t>((t-static_cast<double>(t_sec)) * 1e9);
+    return rclcpp::Time{t_sec, t_nsec, RCL_ROS_TIME};
+  }
+
+  bool find_nearest_non_static_model(
+      const std::vector<gazebo::physics::ModelPtr>& models,
+      gazebo::physics::ModelPtr& nearest_model) const
+  {
+    double nearest_dist = 1e6;
+    bool found = false;
+
+    for (const auto& m : models)
+    {
+      if (!m || m->IsStatic() || m->GetName() == _model->GetName())
+        continue;
+      
+      const double dist =
+          m->WorldPose().Pos().Distance(_model->WorldPose().Pos());
+      if (dist < nearest_dist)
+      {
+        nearest_dist = dist;
+        nearest_model = m;
+        found = true;
+      }
+    }
+    return found;
+  }
+
+  bool get_robot_delivery_content(
```

Comment:
I find the name of this function to be a bit ambiguous. Could we rename to `get_payload_model()` or something similar? The argument can be named `payload_model` as well.